agreement/credentialArrivalHistory.go

```diff
@@ -0,0 +1,83 @@
+// Copyright (C) 2019-2023 Algorand, Inc.
+// This file is part of go-algorand
+//
+// go-algorand is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// go-algorand is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package agreement
+
+import (
+	"sort"
+	"time"
+)
+
+// credentialArrivalHistory maintains a circular buffer of time.Duration samples.
+type credentialArrivalHistory struct {
+	history  []time.Duration
+	writePtr int
+	full     bool
+}
+
+func makeCredentialArrivalHistory(size int) credentialArrivalHistory {
+	if size < 0 {
+		panic("can't create CredentialArrivalHistory with negative size")
+	}
+	history := credentialArrivalHistory{history: make([]time.Duration, size)}
+	history.reset()
+	return history
+}
+
+// store saves a new sample into the circular buffer.
+// If the buffer is full, it overwrites the oldest sample.
+func (history *credentialArrivalHistory) store(sample time.Duration) {
+	if len(history.history) == 0 {
+		return
+	}
+
+	history.history[history.writePtr] = sample
+	history.writePtr++
+	if history.writePtr == len(history.history) {
+		history.full = true
+		history.writePtr = 0
+	}
+}
+
+// reset marks the history buffer as empty
+func (history *credentialArrivalHistory) reset() {
+	history.writePtr = 0
+	history.full = false
+}
+
+// isFull checks if the circular buffer has been fully populated at least once.
```

Comment:
I suppose this is a bit silly, but if you want the 37th best of 40 and you're trying to avoid answering too low, you could start answering once you have 37 samples, and return the worst.  We could call this `isReady`.